[SPARK-5521] PCA wrapper for easy transform vectors

catap opened the pull request:
I implement a simple PCA wrapper for easy transform of vectors by PCA for example LabeledPoint or another complicated structure.

Example of usage:

```
  import org.apache.spark.mllib.regression.LinearRegressionWithSGD
  import org.apache.spark.mllib.regression.LabeledPoint
  import org.apache.spark.mllib.linalg.Vectors
  import org.apache.spark.mllib.feature.PCA

  val data = sc.textFile("data/mllib/ridge-data/lpsa.data").map { line =>
    val parts = line.split(',')
    LabeledPoint(parts(0).toDouble, Vectors.dense(parts(1).split(' ').map(_.toDouble)))
  }.cache()

  val splits = data.randomSplit(Array(0.6, 0.4), seed = 11L)
  val training = splits(0).cache()
  val test = splits(1)

  val pca = PCA.create(training.first().features.size/2, data.map(_.features))
  val training_pca = training.map(p => p.copy(features = pca.transform(p.features)))
  val test_pca = test.map(p => p.copy(features = pca.transform(p.features)))

  val numIterations = 100
  val model = LinearRegressionWithSGD.train(training, numIterations)
  val model_pca = LinearRegressionWithSGD.train(training_pca, numIterations)

  val valuesAndPreds = test.map { point =>
    val score = model.predict(point.features)
    (score, point.label)
  }

  val valuesAndPreds_pca = test_pca.map { point =>
    val score = model_pca.predict(point.features)
    (score, point.label)
  }

  val MSE = valuesAndPreds.map{case(v, p) => math.pow((v - p), 2)}.mean()
  val MSE_pca = valuesAndPreds_pca.map{case(v, p) => math.pow((v - p), 2)}.mean()

  println("Mean Squared Error = " + MSE)
  println("PCA Mean Squared Error = " + MSE_pca)
```

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
@catap This is nice to have. Could you follow the steps in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark for contributing to Spark? For example, you need to create a JIRA (and get assigned) and put the JIRA number in the PR title.

For the public APIs, please follow other transformers under `mllib.feature`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,79 @@
+ * Transformer a vector use PCA
+ */
+
+object PCA {
```
Please use class, setters, and the`fit` method as in other feature transformers under `mllib.feature`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,79 @@
+   * @return transformed vector.
+   */
+  override def transform(vector: Vector): Vector = {
+    val mat = Matrices.dense(1, vector.size, vector.toArray).multiply(pc)
```
`BLAS.gemv` should work. We might miss the implementation of dense matrix * sparse vector, but it should be easy to add.

catap replied:
@mengxr, `Matrix.multiply` is a simple wrapper over `BLAS.gemv`. I think is a wrong way to duplicate this wrapper here, isn't it?

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,63 @@
+package org.apache.spark.mllib.feature
```
Add Apache copyright header.

mengxr commented:
ok to test

SparkQA commented:
[Test build #26570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26570/consoleFull) for   PR 4304 at commit [`ba86afb`](https://github.com/apache/spark/commit/ba86afbb6fa76c2e4888c86f15b0ba3382db4af0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26570/consoleFull) for   PR 4304 at commit [`ba86afb`](https://github.com/apache/spark/commit/ba86afbb6fa76c2e4888c86f15b0ba3382db4af0).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26570/
Test FAILed.

SparkQA commented:
[Test build #26613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26613/consoleFull) for   PR 4304 at commit [`d2a22cf`](https://github.com/apache/spark/commit/d2a22cf82041eb16492eb5f6202717d9558ba0ee).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26613/consoleFull) for   PR 4304 at commit [`d2a22cf`](https://github.com/apache/spark/commit/d2a22cf82041eb16492eb5f6202717d9558ba0ee).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26613/
Test PASSed.

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  test("Correct k size: k = 1") {
+    new PCA(1).fit(dataRDD)
```
What is the test? For the example matrix, we should compute its expected output (in MATLAB or in Python) validate the result.

catap replied:
It's a stupid test, yes. I agree with you :)

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,76 @@
+    assert(pca_transform.deep == mat_multiply.deep)
+  }
+
+  def test_wrong_k(k: Int) = {
```
We don't use underscore in Scala's method names. And to assert failures, we use `withClue` and `intercept`. See

https://github.com/apache/spark/blob/84324fbcb987db6e10e435f463eacace1bae43e2/mllib/src/test/scala/org/apache/spark/mllib/stat/MultivariateOnlineSummarizerSuite.scala#L44

SparkQA commented:
[Test build #26632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26632/consoleFull) for   PR 4304 at commit [`8b29946`](https://github.com/apache/spark/commit/8b29946d0b1aabfcb393f71f9de6202cee51a39d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26632/consoleFull) for   PR 4304 at commit [`8b29946`](https://github.com/apache/spark/commit/8b29946d0b1aabfcb393f71f9de6202cee51a39d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26632/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented on docs/mllib-dimensionality-reduction.md:
```diff
@@ -157,6 +157,23 @@ val pc: Matrix = mat.computePrincipalComponents(10) // Principal components are
 val projected: RowMatrix = mat.multiply(pc)
 {% endhighlight %}
+
+The following code demonstrates how to compute principal components on a `Vector`
```
`a Vector` -> `source vectors`

mengxr commented on docs/mllib-dimensionality-reduction.md:
```diff
@@ -157,6 +157,23 @@ val pc: Matrix = mat.computePrincipalComponents(10) // Principal components are
 {% endhighlight %}
+
+The following code demonstrates how to compute principal components on a `Vector`
+and use them to project the vectors into a low-dimensional space keep vectors label
```
`while keeping associated labels:`

mengxr commented on docs/mllib-feature-extraction.md:
```diff
@@ -370,3 +370,58 @@ data2 = labels.zip(normalizer2.transform(features))
+
+## PCA
+
+PCA transform is a easy way to reduction size of vectors.
```
Please check the grammar of the entire section. You can use the grammar checker that comes with Word or other editors.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+import org.apache.spark.rdd.RDD
+
+/**
+ * Transformer a vector use PCA
```
`Transformer` is not a verb. `A feature transformer that projects vectors to a low-dimensional space using PCA.`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+/**
+ * Transformer a vector use PCA
+ *
+ * @param k count of principal components
```
`num of principal components` (This number is requested by user, not a result count.)

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+  require(k >= 1)
+
+  /**
+   * Compute a Principal Components from RDD[Vector]
```
`Compute a Principal Components` -> `Computes a [[PCAModel] that contains the principal components of the input vectors.`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+      case dm: DenseMatrix =>
+        dm
+      /*
+       * Convert Sparse Vector to Dense Vector.
```
`Convert sparse matrix to dense.` (Do not capitalize the first letter for normal words.)

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+       * Convert Sparse Vector to Dense Vector.
+       *
+       * Following code is for possible compatibility.
+       * RowMatrix.computePrincipalComponents is always returned a Dense Vector
```
`RowMatrix.computePrincipalComponents always returns a dense matrix. The following code is a safeguard.`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  /**
+   * Compute a Principal Components from JavaRDD[Vector]
```
See my previous comment on `fit`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+   *
+   * @param sources source vectors
+   */
+  def fit(k: Int, sources: JavaRDD[Vector]): PCAModel =
```
Does it compile?

catap replied:
Yes, it does.
Why shouldn't it compile?

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+}
+
+/**
+ * This class is wrapper of RowMatrix.computePrincipalComponents and Matrix.multiply
```
This is implementation detail but not user-facing doc. `Model fitted by [[PCA]] that can project vectors to a low-dimensional space spanned by principal components.`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,111 @@
+   * @return transformed vector.
+   */
+  override def transform(vector: Vector): Vector = {
+    val mat = vector match {
```
`pc.transpose.multiply(vector)`

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,47 @@
+import org.apache.spark.mllib.linalg.Vectors
+import org.apache.spark.mllib.linalg.distributed.RowMatrix
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.scalatest.FunSuite
```
organize imports (https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports)

mengxr commented:
ok to test

SparkQA commented:
[Test build #27291 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27291/consoleFull) for   PR 4304 at commit [`8b29946`](https://github.com/apache/spark/commit/8b29946d0b1aabfcb393f71f9de6202cee51a39d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27291 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27291/consoleFull) for   PR 4304 at commit [`8b29946`](https://github.com/apache/spark/commit/8b29946d0b1aabfcb393f71f9de6202cee51a39d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27291/
Test FAILed.

SparkQA commented:
[Test build #27433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27433/consoleFull) for   PR 4304 at commit [`4d63fd1`](https://github.com/apache/spark/commit/4d63fd18ce937ec973295214d304a32af1bae064).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27433/consoleFull) for   PR 4304 at commit [`4d63fd1`](https://github.com/apache/spark/commit/4d63fd18ce937ec973295214d304a32af1bae064).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27433/
Test FAILed.

SparkQA commented:
[Test build #27434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27434/consoleFull) for   PR 4304 at commit [`69ba57c`](https://github.com/apache/spark/commit/69ba57c411b361475122eb0aec59ab52490d3603).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27434/consoleFull) for   PR 4304 at commit [`69ba57c`](https://github.com/apache/spark/commit/69ba57c411b361475122eb0aec59ab52490d3603).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27434/
Test FAILed.

SparkQA commented:
[Test build #27436 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27436/consoleFull) for   PR 4304 at commit [`25d77b3`](https://github.com/apache/spark/commit/25d77b30c2ec49e84eb1f5f7f90318edd466ea68).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27436 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27436/consoleFull) for   PR 4304 at commit [`25d77b3`](https://github.com/apache/spark/commit/25d77b30c2ec49e84eb1f5f7f90318edd466ea68).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27436/
Test FAILed.

SparkQA commented:
[Test build #27437 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27437/consoleFull) for   PR 4304 at commit [`d7f6725`](https://github.com/apache/spark/commit/d7f6725f22943130d01a9dcb8b7f5688ff708834).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27437 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27437/consoleFull) for   PR 4304 at commit [`d7f6725`](https://github.com/apache/spark/commit/d7f6725f22943130d01a9dcb8b7f5688ff708834).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27437/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

jkbradley commented:
@catap  Apologies for the delayed review!  I'll make a pass through it now.  It would be great to get this into the Spark 1.4 release (code cutoff this Friday).

jkbradley commented on docs/mllib-feature-extraction.md:
```diff
@@ -375,3 +375,58 @@ data2 = labels.zip(normalizer2.transform(features))
+println("PCA Mean Squared Error = " + MSE_pca)
+{% endhighlight %}
+</div>
+</div>
```
missing newline

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+/**
+ * A feature transformer that projects vectors to a low-dimensional space using PCA.
+ *
+ * @param k num of principal components
```
"num" --> "number"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+ *
+ * @param k num of principal components
+ */
+
```
remove newline

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+   * @param sources source vectors
+   */
+  def fit(sources: RDD[Vector]): PCAModel = {
+    require(k <= sources.first().size)
```
Please add error message

jkbradley replied:
What happens when sources.count() < k?  I'm not sure if PCA is robust (returning some all-zero components) or if it dies.  Could you please either verify that PCA is robust, or add a check here?  Thanks!

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+    val pc = mat.computePrincipalComponents(k) match {
+      case dm: DenseMatrix =>
+        dm
+      /*
```
scala style: move comment 1 line down to line 53 (within the block handling SparseMatrix)

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Model fitted by [[PCA]] that can project vectors to a low-dimensional space space using PCA.
+ *
+ * @param k num of principal components.
```
"num" --> "number"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+ * Model fitted by [[PCA]] that can project vectors to a low-dimensional space space using PCA.
+ *
+ * @param k num of principal components.
+ * @param pc a principal components Matrix
```
Augment doc: "Each column is one principal component."

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+  /**
+   * Transform a vector by computed Principal Components.
+   *
+   * @param vector vector to be transformed.
```
Augment doc: "Vector must be the same length as the source vectors given to PCA.fit()"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+   * Transform a vector by computed Principal Components.
+   *
+   * @param vector vector to be transformed.
+   * @return transformed vector.
```
Augment doc: "Vector will be of length k"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,109 @@
+        pc.transpose.multiply(dv)
+      case sv: SparseVector =>
+        /* SparseVector -> single row SparseMatrix */
+        val rowIndices = new Array[Int](sv.indices.length)
```
Would this be easier?
```
Matrices.sparse(sv.size, 1, Array(0, sv.indices.length), sv.indices, sv.values).transpose()
```

catap replied:
Thanks!

Here I made a matrix transpose by hands :)

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,48 @@
+    val pca_transform = pca.transform(dataRDD).collect()
+    val mat_multiply = mat.multiply(pc).rows.collect()
+
+    assert(pca_transform.deep == mat_multiply.deep)
```
Convert these to sets first since the order is not guaranteed.

Also, does it work to use the operator ```===``` instead of ```==```?  That's better in general for handling types.

jkbradley commented on docs/mllib-dimensionality-reduction.md:
```diff
@@ -157,6 +157,23 @@ val pc: Matrix = mat.computePrincipalComponents(10) // Principal components are
+import org.apache.spark.mllib.regression.LabeledPoint
+import org.apache.spark.mllib.feature.PCA
+
+val data: LabeledPoint = ...
```
This should be an RDD

jkbradley commented on docs/mllib-dimensionality-reduction.md:
```diff
@@ -157,6 +157,23 @@ val pc: Matrix = mat.computePrincipalComponents(10) // Principal components are
+// Compute the top 10 principal components.
+val pca = new PCA(10).fit(data.map(_.features))
+
+// Project vectors to the linear space spanned by the top 10 principal components keep vectors label
```
"components keep vectors label" --> "components, keeping the label"

jkbradley commented:
@catap I think that's all for now.  Most of those are small comments.  Thanks in advance!

catap commented:
@jkbradley, I will fix your comments today evening or tomorrow morning. Thanks!

SparkQA commented:
[Test build #766 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/766/consoleFull) for   PR 4304 at commit [`d7f6725`](https://github.com/apache/spark/commit/d7f6725f22943130d01a9dcb8b7f5688ff708834).
 * This patch **does not merge cleanly**.

jkbradley commented:
@catap I'm sorry about the delay!  I'm not sure why testing wasn't triggered by your last update, but I just saw it now because of that.  Could you please fix the current merge issues (probably best to rebase off of the current master)?  Thank you!

SparkQA commented:
[Test build #766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/766/consoleFull) for   PR 4304 at commit [`d7f6725`](https://github.com/apache/spark/commit/d7f6725f22943130d01a9dcb8b7f5688ff708834).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

catap commented:
@jkbradley it's funny, but commit had a conflict ;)

Fix it!

jkbradley commented on docs/mllib-dimensionality-reduction.md:
```diff
@@ -157,6 +157,23 @@ val pc: Matrix = mat.computePrincipalComponents(10) // Principal components are
+// Compute the top 10 principal components.
+val pca = new PCA(10).fit(data.map(_.features))
+
+// Project vectors to the linear space spanned by the top 10 principal components, keeping the labe
```
typo: "labe" --> "label"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,97 @@
+   */
+  def fit(sources: RDD[Vector]): PCAModel = {
+    require(k <= sources.first().size,
+      s"source vector size is ${sources.first().size} must be a great of k that is $k")
```
```must be a great of k that is $k``` should be ```must be greater than k=$k```

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,97 @@
+   * @param sources source vectors
+   */
+  def fit(k: Int, sources: JavaRDD[Vector]): PCAModel =
+    fit(k, sources.rdd)
```
put on line above:
```
  def fit(k: Int, sources: JavaRDD[Vector]): PCAModel = fit(k, sources.rdd)
```

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -0,0 +1,97 @@
+ * @param k number of principal components.
+ * @param pc a principal components Matrix. Each column is one principal component.
+ */
+class PCAModel private[mllib](val k: Int, val pc: DenseMatrix) extends VectorTransformer {
```
space needed: ```private[mllib] (val k: Int```

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -0,0 +1,48 @@
+    val pca_transform = pca.transform(dataRDD).collect()
+    val mat_multiply = mat.multiply(pc).rows.collect()
+
+    assert(pca_transform.deep === mat_multiply.deep)
```
Convert these to sets first since the order is not guaranteed.

jkbradley commented:
What happens when sources.count() < k? I'm not sure if PCA is robust (returning some all-zero components) or if it dies. Could you please either verify that PCA is robust, or add a check here? Thanks!

jkbradley commented:
There are a few remaining items I commented on just now, but hopefully that will be it.  Thanks!

catap committed:
[SPARK-5521] PCA wrapper for easy transform vectors

I implement a simple PCA wrapper for easy transform of vectors by PCA for example LabeledPoint or another complicated structure.

Example of usage:
```
  import org.apache.spark.mllib.regression.LinearRegressionWithSGD
  import org.apache.spark.mllib.regression.LabeledPoint
  import org.apache.spark.mllib.linalg.Vectors
  import org.apache.spark.mllib.feature.PCA

  val data = sc.textFile("data/mllib/ridge-data/lpsa.data").map { line =>
    val parts = line.split(',')
    LabeledPoint(parts(0).toDouble, Vectors.dense(parts(1).split(' ').map(_.toDouble)))
  }.cache()

  val splits = data.randomSplit(Array(0.6, 0.4), seed = 11L)
  val training = splits(0).cache()
  val test = splits(1)

  val pca = new PCA(training.first().features.size/2).fit(data.map(_.features))
  val training_pca = training.map(p => p.copy(features = pca.transform(p.features)))
  val test_pca = test.map(p => p.copy(features = pca.transform(p.features)))

  val numIterations = 100
  val model = LinearRegressionWithSGD.train(training, numIterations)
  val model_pca = LinearRegressionWithSGD.train(training_pca, numIterations)

  val valuesAndPreds = test.map { point =>
    val score = model.predict(point.features)
    (score, point.label)
  }

  val valuesAndPreds_pca = test_pca.map { point =>
    val score = model_pca.predict(point.features)
    (score, point.label)
  }

  val MSE = valuesAndPreds.map{case(v, p) => math.pow((v - p), 2)}.mean()
  val MSE_pca = valuesAndPreds_pca.map{case(v, p) => math.pow((v - p), 2)}.mean()

  println("Mean Squared Error = " + MSE)
  println("PCA Mean Squared Error = " + MSE_pca)
```

jkbradley commented:
Jenkins test this please

jkbradley commented:
@catap Thanks for the update.  In the future, when you update a PR, can you please try to push only the new commit?  It looks like you're rebasing or squashing your commit each time, which makes it hard to see updates you make after each pass of reviewing.  Thanks!

I think this will be ready once tests pass.

SparkQA commented:
[Test build #770 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/770/consoleFull) for   PR 4304 at commit [`5a16f20`](https://github.com/apache/spark/commit/5a16f20fb35fb221d70917ffbfac7d5ca7a7d7d3).

SparkQA commented:
[Test build #770 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/770/consoleFull) for   PR 4304 at commit [`5a16f20`](https://github.com/apache/spark/commit/5a16f20fb35fb221d70917ffbfac7d5ca7a7d7d3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "public class " + className + extendsText + " implements java.io.Serializable `
  * `abstract class ShuffleHandle(val shuffleId: Int) extends Serializable `
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`
  * `abstract class Evaluator extends Params `
  * `abstract class PipelineStage extends Params with Logging `
  * `class BinaryClassificationEvaluator extends Evaluator with HasRawPredictionCol with HasLabelCol `
  * `trait LDAOptimizer `
  * `class EMLDAOptimizer extends LDAOptimizer `
  * `class OnlineLDAOptimizer extends LDAOptimizer `
  * `class PCA(val k: Int) `
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `
  * `class BinaryClassificationEvaluator(JavaEvaluator, HasLabelCol, HasRawPredictionCol):`
  * `class HasRawPredictionCol(Params):`
  * `class Evaluator(object):`
  * `class ParamGridBuilder(object):`
  * `class JavaEvaluator(Evaluator, JavaWrapper):`
  * `class DataFrameStatFunctions(object):`
  * `abstract class Dialect `
  * `  class DialectException(msg: String, cause: Throwable) extends Exception(msg, cause)`
  * `case class Not(child: Expression) extends UnaryExpression with Predicate with ExpectsInputTypes `
  * `case class And(left: Expression, right: Expression)`
  * `case class Or(left: Expression, right: Expression)`
  * `abstract class BinaryComparison extends BinaryExpression with Predicate `
  * `trait StringRegexExpression extends ExpectsInputTypes `
  * `trait CaseConversionExpression extends ExpectsInputTypes `
  * `trait StringComparison extends ExpectsInputTypes `
  * `case class Substring(str: Expression, pos: Expression, len: Expression)`
  * `case class HiveDatabase(`
  * `abstract class TableType `
  * `case class HiveStorageDescriptor(`
  * `case class HivePartition(`
  * `case class HiveColumn(name: String, hiveType: String, comment: String)`
  * `case class HiveTable(`
  * `trait ClientInterface `
  * `class ClientWrapper(`
  * `class IsolatedClientLoader(`
  * `protected trait ReflectionMagic `
  * `  protected implicit class InstanceMagic(a: Any) `
  * `  protected implicit class StaticMagic(c: Class[_]) `

jkbradley commented:
@catap Could you please fix the Scala style test?
```
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:80: File line length exceeds 100 characters
```

catap commented:
@jkbradley np

catap committed:
[SPARK-5521] fix scala style

catap commented:
@jkbradley, yes I use rebase to create one functional commit. If it works for you, I can push base commit and different commit for fixing.

SparkQA commented:
[Test build #773 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/773/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).

SparkQA commented:
[Test build #773 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/773/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

jkbradley commented:
Jenkins test this please

SparkQA commented:
[Test build #776 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/776/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).

SparkQA commented:
[Test build #776 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/776/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

jkbradley committed:
Small updates: removed k from Java-friendly PCA fit().  In PCASuite, converted results to set for comparison. Added an error message for bad k in PCA.

jkbradley commented:
@catap I noticed there is still an unaddressed comment or two.  I'm going to send a PR to your branch with a few updates.

SparkQA commented:
[Test build #784 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/784/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).

SparkQA commented:
[Test build #784 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/784/consoleFull) for   PR 4304 at commit [`1e0e518`](https://github.com/apache/spark/commit/1e0e518f6687a39ac6b4188b3dfcc94b97d19197).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `

jkbradley commented:
@catap It looks like this doesn't merge cleanly now.  (There are lots of PRs being merged nowadays.)  Could you please rebase and fix the issues?  (It's OK to rebase to fix bad merges.)  Thanks!

catap commented:
@jkbradley, done!

SparkQA commented:
[Test build #795 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/795/consoleFull) for   PR 4304 at commit [`501bcd9`](https://github.com/apache/spark/commit/501bcd9f2daf09a941334a2a1198b8d4f1efd7cf).

SparkQA commented:
[Test build #795 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/795/consoleFull) for   PR 4304 at commit [`501bcd9`](https://github.com/apache/spark/commit/501bcd9f2daf09a941334a2a1198b8d4f1efd7cf).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
Spurious failure.  Retesting...

SparkQA commented:
[Test build #796 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/796/consoleFull) for   PR 4304 at commit [`501bcd9`](https://github.com/apache/spark/commit/501bcd9f2daf09a941334a2a1198b8d4f1efd7cf).

SparkQA commented:
[Test build #796 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/796/consoleFull) for   PR 4304 at commit [`501bcd9`](https://github.com/apache/spark/commit/501bcd9f2daf09a941334a2a1198b8d4f1efd7cf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PCA(val k: Int) `
  * `class HasCheckpointInterval(Params):`
  * `class ALS(JavaEstimator, HasCheckpointInterval, HasMaxIter, HasPredictionCol, HasRegParam, HasSeed):`
  * `class ALSModel(JavaModel):`

jkbradley commented:
@catap LGTM merging into master and branch-1.4.  Thanks very much!
